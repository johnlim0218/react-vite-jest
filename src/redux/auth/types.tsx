export interface AuthState {
  authKey: string | null;
  employeeData: EmployeeData | null;
  employeeCompanyName: string;
}

export interface EmployeeData {
  employeeId: string;
  name: string;
  roles: any;
}