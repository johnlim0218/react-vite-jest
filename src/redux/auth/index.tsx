import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { CompanyCode } from '@utils/mapping/company';

interface AuthState {
  authKey: string | null;
  employeeData: EmployeeData | null;
  employeeCompanyName: string;
}

interface EmployeeData {
  employeeId: string;
  name: string;
  roles: any;
}

export const initialAuthKey = () => {
  return localStorage.getItem('authKey');
};

const initialState: AuthState = {
  authKey: initialAuthKey(),
  employeeData: { employeeId: '', name: '', roles: '' },
  employeeCompanyName: ''
}

const authSlice = createSlice({
  name: 'auth',
  initialState,
  reducers: {
    setAuthKey: (state, action: PayloadAction<string>) => {
      state.authKey = action.payload;
      localStorage.setItem('authKey', action.payload);
    },
    setEmployeeData: (state, action: PayloadAction<EmployeeData>) => {
      state.employeeData = action.payload;
    },
    setEmployeeCompanyData: (state, action: PayloadAction<number>) => {
      state.employeeCompanyName = CompanyCode[action.payload];
    },
    clearAuth: (state) => {
      state.authKey = null;
      state.employeeData = null;
      localStorage.removeItem('authKey');
    }
  }
});

export const { 
  setAuthKey,
  setEmployeeData,
  setEmployeeCompanyData,
  clearAuth 
} = authSlice.actions;

export default authSlice.reducer