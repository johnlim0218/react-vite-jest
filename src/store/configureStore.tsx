import { AnyAction, configureStore, Dispatch, MiddlewareAPI } from '@reduxjs/toolkit';
// import { composeWithDevTools } from "redux-devtools-extension";
import rootReducer from '../redux/rootReducer';

const loggerMiddleware = (store: MiddlewareAPI<Dispatch, any>) => (next: Dispatch<AnyAction>) => (action: any) => {
  const result = next(action);
  return result;
};

const middlewares = [loggerMiddleware];

const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) => 
    getDefaultMiddleware()
      .concat(
        middlewares
      ),
  devTools: process.env.NODE_ENV !== 'production'
})

export type AppDispatch = typeof store.dispatch;

export default store;
