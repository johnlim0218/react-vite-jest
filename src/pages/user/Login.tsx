import * as React from 'react';
import { useDispatch } from 'react-redux';
import { testAction } from '@redux/auth';

function LoginPage() {
  const dispatch = useDispatch();

  const onClickButton = () => {
    dispatch(testAction());
  }
  return (
    <div>
      <div>로그인</div>
      <button onClick={onClickButton}>버튼</button>
    </div>
  )

}

export default LoginPage;