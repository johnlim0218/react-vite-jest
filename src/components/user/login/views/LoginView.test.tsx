import * as React from 'react';
import { fireEvent, render, RenderResult, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import * as authActions from '@redux/auth';
import store from '@store/configureStore';
import Login from '..';

describe('LoginView', () => {
  let utils: RenderResult;
  const mockStore = configureMockStore();

  // 가짜 스토어
  let store = mockStore({
    auth: {
      employeeData: { employeeId: '', name: '', roles: '' }
    }
  });

  it('로그인 뷰 스냅샷', () => {
    utils = render(
      <Provider store={store}>
        <Login />
      </Provider>
    );
    expect(utils.container).toMatchSnapshot();
  });

  it('로그인 타이틀', () => {
    utils = render(
      <Provider store={store}>
        <Login />
      </Provider>
    );
    const titleElement = screen.getByText('로그인');
    expect(titleElement).toBeInTheDocument();
  });

  it('로그인 onSubmit - "비밀번호가 틀렸습니다."', () => {
    // onSubmit 함수 테스트 여기서 하는건가?
    render(
      <Provider store={store}>
        <Login />
      </Provider>
    );

    const employeeIdInput = screen.getByLabelText('employeeId');
    fireEvent.change(employeeIdInput, { target: { value: '1234567' } });
    expect((employeeIdInput as HTMLInputElement).value).toBe('1234567');

    const passwordInput = screen.getByLabelText('password');
    fireEvent.change(passwordInput, { target: { value: '2' } });
    expect((passwordInput as HTMLInputElement).value).toBe('2');

    const button = screen.getByRole('button');
    fireEvent.click(button);

    const wrongPasswordMessage = screen.getByText('비밀번호가 틀렸습니다.');
    expect(wrongPasswordMessage).toHaveTextContent('비밀번호가 틀렸습니다.');
  });

  it('로그인 onSubmit - setEmployeeData()', () => {
    render(
      <Provider store={store}>
        <Login />
      </Provider>
    );

    const employeeIdInput = screen.getByLabelText('employeeId');
    fireEvent.change(employeeIdInput, { target: { value: '1234567' } });
    expect((employeeIdInput as HTMLInputElement).value).toBe('1234567');

    const passwordInput = screen.getByLabelText('password');
    fireEvent.change(passwordInput, { target: { value: '1' } });
    expect((passwordInput as HTMLInputElement).value).toBe('1');

    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(store.getActions()[0]).toEqual(
      authActions.setEmployeeData({
        employeeId: '1234567',
        name: '관우림',
        roles: 'EMPLOYEE'
      })
    );
    // store.dispatch(authActions.setEmployeeData({
    //   employeeId: '1234567',
    //   name: '관우림',
    //   roles: 'EMPLOYEE'
    // }));

    // expect(store.getActions()).toEqual([
    //   {
    //     type: 'auth/setEmployeeData',
    //     payload: {
    //       employeeId: '1234567',
    //       name: '관우림',
    //       roles: 'EMPLOYEE'
    //     }
    //   }
    // ]);
  });
});
