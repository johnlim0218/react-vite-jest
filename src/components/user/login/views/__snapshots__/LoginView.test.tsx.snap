// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginView 로그인 뷰 스냅샷 1`] = `
<div>
  <div>
    <div>
      로그인
    </div>
    <form>
      <div>
        <p>
          사원번호
        </p>
        <input
          aria-label="employeeId"
          name="employeeId"
          value=""
        />
      </div>
      <div>
        <p>
          비밀번호
        </p>
        <input
          aria-label="password"
          name="password"
          type="password"
          value=""
        />
      </div>
      <button
        type="submit"
      >
        버튼
      </button>
    </form>
  </div>
</div>
`;
